### PR TITLE
improving api/components/suggestions for better UI usability

### PR DESCRIPTION
**Pull request by bartfastiel**

### CODE REVIEW

File: server/sonar-server/src/test/java/org/sonar/server/component/ws/SuggestionsActionTest.java [hunk 1]
```diff
@@ -109,6 +110,110 @@ public void define_suggestions_action() {
...
+  @Test
+  public void suggestions_without_query_should_contain_recently_browsed() throws Exception {
+    ComponentDto project = db.components().insertComponent(newPrivateProjectDto(organization));
+
+    componentIndexer.indexOnStartup(null);
+    authorizationIndexerTester.allowOnlyAnyone(project);
+
+    SuggestionsWsResponse response = actionTester.newRequest()
+      .setMethod("POST")
+      .setParam(PARAM_RECENTLY_BROWSED, Stream.of(project.getKey()).collect(joining(",")))
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/component/ws/SuggestionsActionTest.java]: the stream is useless here, just put the project key

File: server/sonar-server/src/test/java/org/sonar/server/component/ws/SuggestionsActionTest.java [hunk 2]
```diff
@@ -109,6 +110,110 @@ public void define_suggestions_action() {
...
+    ComponentDto project4 = db.components().insertComponent(newPrivateProjectDto(organization).setName("Delta"));
+    doReturn(asList(project4, project2)).when(favoriteFinder).list();
+
+    componentIndexer.indexOnStartup(null);
+    authorizationIndexerTester.allowOnlyAnyone(project1);
+    authorizationIndexerTester.allowOnlyAnyone(project2);
+    authorizationIndexerTester.allowOnlyAnyone(project3);
+    authorizationIndexerTester.allowOnlyAnyone(project4);
+
+    SuggestionsWsResponse response = actionTester.newRequest()
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/component/ws/SuggestionsActionTest.java]: remark: in many places, WsActionTester is named ws in the test. For consistency, I suggest to use the same name

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java [hunk 3]
```diff
@@ -131,75 +134,141 @@ public void handle(Request wsRequest, Response wsResponse) throws Exception {
     }
     Set<String> favoriteKeys = favoriteFinder.list().stream().map(ComponentDto::getKey).collect(Collectors.toSet());
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: recentlyBrowsedKeys could be extracted to a specific method:
```
Set<String> recentlyBrowsedKeys = getRecentlyBrowsedKeys(wsRequest);
```

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java [hunk 4]
```diff
@@ -131,75 +134,141 @@ public void handle(Request wsRequest, Response wsResponse) throws Exception {
...
 
+    List<String> qualifiers = getQualifiers(more);
+    SuggestionsWsResponse searchWsResponse;
+    if (query == null) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: extract to method `searchSuggestions`

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java [hunk 5]
```diff
@@ -131,75 +134,141 @@ public void handle(Request wsRequest, Response wsResponse) throws Exception {
...
+    }
+    writeProtobuf(searchWsResponse, wsRequest, wsResponse);
+  }
+
+  private SuggestionsWsResponse suggestionsWithoutSearch(@Nullable String more, Set<String> recentlyBrowsedKeys, Set<String> favoriteKeys, List<String> qualifiers) {
+    if (favoriteKeys.isEmpty() && recentlyBrowsedKeys.isEmpty()) {
+      return newBuilder().build();
+    }
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Map<String, List<ComponentDto>> favorites = dbClient.componentDao().selectByKeys(dbSession, favoriteKeys)
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: use `MoreCollectors.index`

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java [hunk 6]
```diff
@@ -131,75 +134,141 @@ public void handle(Request wsRequest, Response wsResponse) throws Exception {
...
+        .stream().collect(Collectors.groupingBy(ComponentDto::qualifier));
+      Map<String, List<ComponentDto>> recentlyBrowsed = dbClient.componentDao().selectByKeys(dbSession, recentlyBrowsedKeys)
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: use `MoreCollectors.index`

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java [hunk 7]
```diff
@@ -131,75 +134,141 @@ public void handle(Request wsRequest, Response wsResponse) throws Exception {
...
+
+      Stream<ComponentDto> componentDtoStream = Stream.concat(favorites.values().stream(), recentlyBrowsed.values().stream())
+        .flatMap(Collection::stream);
+      return buildResponse(recentlyBrowsedKeys, favoriteKeys, componentsPerQualifiers, dbSession, componentDtoStream).build();
+    }
+  }
+
+  private SuggestionsWsResponse suggestionsFromSearch(String query, @Nullable String more, Set<String> recentlyBrowsedKeys, Set<String> favoriteKeys,
+    List<String> qualifiers) {
+    ComponentIndexResults componentsPerQualifiers;
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: set it when initialized

File: server/sonar-server/src/test/java/org/sonar/server/component/ws/SuggestionsActionTest.java (hunk 1 above)
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/component/ws/SuggestionsActionTest.java]: Done.

File: server/sonar-server/src/test/java/org/sonar/server/component/ws/SuggestionsActionTest.java (hunk 2 above)
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/component/ws/SuggestionsActionTest.java]: Done.

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java (hunk 3 above)
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: Done (also for `getFavoriteKeys`). Much better now.

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java (hunk 4 above)
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: Done.

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java (hunk 5 above)
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: Done.

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java (hunk 6 above)
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: Done.

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java (hunk 7 above)
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: Done.

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java [hunk 8]
```diff
@@ -115,99 +126,167 @@ public void define(WebService.NewController context) {
...
-      qualifiers = singletonList(SuggestionCategory.getByName(more).getQualifier());
-      queryBuilder.setLimit(EXTENDED_LIMIT);
+      return stream(SuggestionCategory.values()).map(SuggestionCategory::getQualifier).collect(Collectors.toList());
     }
-    queryBuilder.setQualifiers(qualifiers);
-    return searchInIndex(queryBuilder.build());
+    return singletonList(SuggestionCategory.getByName(more).getQualifier());
+  }
+
+  private SuggestionsWsResponse.Builder buildResponse(Set<String> recentlyBrowsedKeys, Set<String> favoriteUuids, ComponentIndexResults componentsPerQualifiers,
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: MINOR Split this 181 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**